pkg/util/wait/delayed_action.go

```diff
@@ -0,0 +1,142 @@
+
+import (
+	"time"
+)
+
+// nowFn is a function for testing
+var nowFn = time.Now
+
+// Interval defines an interval over which a certain number of events are expected.
+type Interval struct {
+	Count    int
+	Interval time.Duration
+}
+
+// NewChannelDelayedAction is an efficient implementation that can throttle actions based on a progressive rate
+// schedule defined by the provided steps. Intended to be used in a for{ select{...} } loop:
+//
+//     delay := NewChannelDelayedAction(Interval{Count: 10, Interval: time.Second})
+//     for {
+//       select {
+//	     case <-delay.After():
+//         delay.Done()
+//         // perform a rate limited action
+//       ...
+//       }
+//       if delay.Run() {
+//	       // perform a rate limited action
+//       }
+//     }
+//
+// If the for loop runs faster than 10 events per second, Run() will run false and the delay channel will have
+// an item when the queue is exhausted. Done() is required to clear the state in the loop. If the rate drops
+// below 10/sec, then no delays will be used. The rate is calculated over the time necessary to accumulate 10
+// events (not an instaneous rate).
+func NewChannelDelayedAction(steps ...Interval) *ChannelDelayedAction {
+	return &ChannelDelayedAction{steps: steps}
+}
+
+// ChannelDelayedAction allows callers to delay expensive actions inside of for-select loops.
+type ChannelDelayedAction struct {
+	active <-chan time.Time
+	t      *time.Timer
+
+	steps    []Interval
+	count    int
+	interval time.Duration
+	last     time.Time
+}
+
+// After is the channel to listen for delayed actions in a for select loop.
+func (q *ChannelDelayedAction) After() <-chan time.Time {
+	return q.active
+}
+
+// Done indicates that the active channel has been read and that subsequent actions should be on
+// a different interval.
+func (q *ChannelDelayedAction) Done() {
+	q.active = nil
+	q.last = time.Time{}
+}
+
+// Run returns true if the action should be performed, and false if it should be run only when After()
+// returns.
+func (q *ChannelDelayedAction) Run() bool {
+	q.count++
+	if q.active != nil {
+		return false
+	}
+	if q.shouldRun() {
+		return true
+	}
+
+	if q.t == nil {
+		q.t = time.NewTimer(q.interval)
+	} else if q.t.Reset(q.interval) {
+		<-q.t.C
+	}
+	q.active = q.t.C
+	return false
+}
+
+// shouldRun returns true if the action should run right away, resetting q.interval and q.count
+// as necessary.
+func (q *ChannelDelayedAction) shouldRun() bool {
+	count := q.count
+
+	// have not accumulated to first threshold, so record start of interval
+	if count < q.steps[0].Count {
+		if q.last.IsZero() {
+			q.last = nowFn()
+		}
+		return true
+	}
+
+	// we're over the first threshold, but we need to have a sufficient rate to be delayed
+	if !q.last.IsZero() {
+		now := nowFn()
+		// calculate our rate and find the first threshold that matches the rate
+		interval := now.Sub(q.last)
+		rate := float32(count) / float32(interval)
+		for i := range q.steps {
```

Comment:
I am still wondering if we can replace this all if we can make watchChan buffered and just do flush when there is not incoming thing in the buffer. But if it is too hard, then probably we can do this time based flush.